Drupal Install on Ubuntu 22.04 (#7338)
linode/docs: docs/guides/websites/cms/drupal/how-to-install-and-configure-drupal-on-ubuntu-22-04/index.md

{{< /note >}}

#### Drupal Project Presence

To confirm the project was scaffolded correctly, verify the presence of key files and directories:

    ls composer.json
    ls web/index.php
    ls -d vendor/

{{< note >}}
If any of these files are missing, Drupal has not been initialized. See the Initialize the Drupal Application Environment section to scaffold the project using Composer.
{{< /note >}}

Contributor-Safe Tips

- Be sure to consider mentioning file permissions if they are relevant to your environment (e.g., `chmod`, `chown`).
- If `vendor/` is missing but `composer.json` is present, run `composer install`.

If at this point all checks pass the infrastructure is ready for Drupal initialization and configuration.

## Create Project Structure

Creating the project structure sets up Drupal for modular development and secure deployment.

This step separates application logic from public content, making updates easier to manage and reducing security risks.

After you run the `composer create-project` command below, your environment will include the core Drupal files and folder structure:

- `composer.json`
- `web/index.php`
- `vendor/`.

> This confirms that setup succeeded, prepares optional configuration scaffolding, and gets the application ready for site installation. *This guide was validated using Drupal 11.1.8 as noted earlier.*

Reviewer: ```suggestion
These files confirm that setup succeeded, prepare optional configuration scaffolding, and get the application ready for site installation. *This guide was validated using Drupal 11.1.8 as noted earlier.*
```

### Create the Drupal Project Structure

> - Run the install command to scaffold (create) the Drupal 11.1.8 structure. Customize the "my_drupal_site" name to fit your needs. For

Reviewer: ```suggestion
- Run the install command to generate the Drupal 11.1.8 structure. Replace "my_drupal_site" with your desired folder name. For 
```

> directory layout details, see [Drupal.org's Directory Structure guide](https://www.drupal.org/docs/getting-started/understanding-drupal/directory-structure).

Reviewer: ```
```

    composer create-project drupal/recommended-project:11.1.8 my_drupal_site

> -Then change to your project folder (removing the <> symbols and replacing with your folder name):

Reviewer: ```suggestion
-Change to your project folder (remove the angle brackets (<>) and use your actual folder name):
```

>     cd <my_drupal_site>

Reviewer: ```suggestion
    cd my-drupal-site
```

- Inside the scaffolded project folder `my_drupal_site`, confirm the environment after installation with:

>     ls composer.json

Reviewer: ```suggestion
    ls composer.json
    ls -ld web/indes.php
    ls -ld vendor/
```

        Result: `composer.json` confirming the metadata file exists

    ls -ld web/index.php
        Response: `web/index.php` confirms the application entry point file exists.

    ls -ld vendor/
        Response:`vendor/` confirming that the `vendor/` directory was created.

If any of these are missing or return errors, installation may have failed or been interrupted. For troubleshooting see [Installing Drupal - Getting Started Guide](https://www.drupal.org/docs/getting-started/installing-drupal).

> **Following Composer Initialization** (scaffolded Drupal Project), confirm your environment against the **Drupal Site Setup checklist**:

Reviewer: ```suggestion
**Following Composer Initialization** (using the recommended Drupal starter template, which places core files like `index.php`, `update.php`, and `.htaccess` in their correct locations), confirm your environment against the **Drupal Site Setup checklist**:
```

|-------------------------------|-------------------------------|------------------------------------------|--------------------------------------------------------------------------|
| Access denied for user        | Wrong username or password    | Double-check credentials in `settings.php` | [Drupal.org: Database Configuration](https://www.drupal.org/docs/drupal-apis/database-api/database-configuration) |
| Unknown database              | Database name typo or missing | Recreate or correct name in config       | [MoldStud: Avoid Common Pitfalls](https://moldstud.com/articles/p-managing-drupal-database-connection-settings-avoid-common-pitfalls) |
| Driver not found              | Incorrect or missing driver   | Use `'driver' => 'mysql'` for MariaDB    | [Drupal.org: Database API Overview](https://www.drupal.org/docs/drupal-apis/database-api/database-configuration) |
| Warning: count() during install | Misconfigured array structure | Ensure `$databases` array is properly nested | [Stack Overflow](https://stackoverflow.com/questions/71596215/how-can-i-set-up-my-drupal-database-correctely) (yes, there is a typo in that title s/b correctly). |

> 1. Optional: Environment Variables

Reviewer: ```suggestion
Optional: Environment Variables
```

{{< note >}}
During installation, you may see a prompt like:

`Do you want to continue?[Y/n]`

This is a standard confirmation step. Type `Y` and press Enter to proceed. (If using a different package manager or install method, the prompt may vary slightly--but the intent is the same: confirm you want to install the listed components.)
{{< /note >}}

**Installation** (Phase 2)

Install the Drupal codebase using Composer. This sets up the recommended project scaffold.

Composer and PHP should already be installed and working. See Phase 1 for environment prep.

| **Step**               | **Purpose**                              | **Command**                                                | **Expected Output**                                   | **If Output Differs**                                      | 🔗 **Further Info** |
|------------------------|------------------------------------------|------------------------------------------------------------|--------------------------------------------------------|------------------------------------------------------------|---------------------|

> | Create Project         | Scaffold Drupal site                     | `composer create-project drupal/recommended-project mysite` | `mysite` folder created with Drupal scaffold           | Rename or delete existing folder before retry              | [Drupal Install Guide](https://www.drupal.org/docs/installing-drupal) |

Reviewer: ```suggestion
| Create Project         | Set up Drupal starter template                     | `composer` | `mysite` folder created with Drupal structure | Rename or delete existing folder before retry              | [Drupal Install Guide](https://www.drupal.org/docs/installing-drupal) |
```

{{< note >}}
These steps help protect your Drupal site from unauthorized access and are strongly recommended for production environments.
{{< /note >}}

2. Enable SSL (Optional)

If deploying Drupal in a production or public-facing environment, configure SSL to encrypt traffic:

- Use *Let's Encrypt* or a self-signed certificate.
- Update your Apache virtual host to include:

<VirtualHost *:443>
    SSLEngine on
    SSLCertificateFile /path/to/cert.pem
    SSLCertificateKeyFile /path/to/key.pem
</VirtualHost>

For local development, SSL is optional. For public sites it is essential.

3. File permissions and `.htaccess` notes

- Ensure the `/web/site/default` directory is writable by the web server during installation:

    sudo chown -R www-data:www-data web/sites/default

- After installation, lock down permissions:

    sudo chmod 444 web/sites/default/settings.php

- Drupal relies on `.htaccess` for security rules like:

    - Preventing access to sensitive files
    - Blocking directory listings
    - Enforcing clean URLs

{{< note >}}
If `.htaccess` rules aren't being applied, double-check `AllowOverride All` is set in your Apache config.
{{< /note >}}

4. Security Checklist

>     - **Database hardened** with `mysql_secure_installation`

Reviewer: ```suggestion
    - **Database hardened** using `mysql_secure_installation`
```

- Avoiding brittle Installs and Opaque Errors

Composer tracks dependencies explicitly, reducing the risk of missing extensions or mismatched versions. Manual installs often fail silently or introduce hard-to-trace errors. Using `composer create-project` ensures a reproducible, contributor-safe environment.

- Encouraging use of Drush for Command-Line Efficiency

Drush streamlines tasks like site installation, cache clearing, and module management. Once installed via Composer, it becomes available in the project’s `/vendor/bin` directory. Run Drush from you project directory:

    ./vendor/bin/drush/status

If you see errors about missing Symfony classes or autoloading failures, double-check that:

    - You're inside the correct project folder (with `composer.json`, `vendor/`, and `web/`).

>     - Drush is intalled locally--not globally or in `/root/vendor`.

Reviewer: ```suggestion
    - Drush is installed locally--not globally or in `/root/vendor`.
```

    - You're not running Drush from outside the project root.

To confirm you're in the right place, look for:

    ls
    # Should include: composer.json, vendor/, web/

### Contributor-Safe References

For official Drush installation guidance, see [Drush on Drupal.org](https://www.drupal.org/docs/develop/development-tools/drush).

You can find additional information on [Drupal's Composer guide](https://www.drupal.org/docs/develop/using-composer/using-composer-with-drupal) for deeper context.

## Conclusion: What Comes Next

> Your Drupal environment is now scaffolded, validated, and ready for customization. From here, you can begin shaping your site:

Reviewer: ```suggestion
Your Drupal environment is now structured, validated, and ready for customization. From here, you can begin shaping your site:
```

- Explore the Official [Drupal Documentation](https://www.drupal.org/documentation) for guidance on modules, configuration, and site building.

> - Get into theme development: create or install themes to control layout and styling. Browse the [Drupal Themes Directory](https://www.drupal.org/project/project_theme) to see hat's possible.

Reviewer: ```suggestion
- Get into theme development: create or install themes to control layout and styling. Browse the [Drupal Themes Directory](https://www.drupal.org/project/project_theme) to see what's possible.
```

## Environment Setup: Preparing to Install Drupal

This section outlines the steps used to verify system components, establish the file structure, and prepare the database--everything Drupal requires prior to installation.

### Infrastructure Verification

Confirm that required components are correctly installed and active to prevent silent failures during setup.

#### Check PHP Version and Extensions

Use the following commands to check what PHP extensions are available:

    php -v
    php -m | grep -E 'gd|mbstring|xml|curl|zip|mysqli|pdo_mysql|opcache'

If any required PHP extensions are missing, install them with:

    sudo apt install php-gd php-mbstring php-xml php-curl php-zip php-mysql php-opcache

Following installation, restart Apache or Nginx to activate the new extensions.

    sudo systemctl restart apache2

Verify they're active:

    php -m | grep -E 'gd|mbstring|xml|curl|zip|mysqli|pdo_mysql|opcache'

Each one should now be listed.

#### Check Composer

If checking for Composer installation yields a `composer` not found error, you can install it with:

    sudo apt update && sudo apt install composer

To verify installation:

    composer --version

> You should see output like `Composer version 2.7.0 or higher.

Author: ```suggestion
You should see output like `Composer version 2.7.0` or higher.
```

Reviewer: This has been fixed.

    sudo chown -R www-data:www-data web

This step prevents permission errors during runtime and ensures that Drupal can safely generate files, manage uploads, and interact with modules.

{{< note >}}
If you're using Nginx or a different web server, replace `www-data` with your actual server user (e.g., `nginx`, `apache`, or a custom service account).
{{< /note >}}

To confirm ownership was updated, run this from the parent directory of `web`:

    ls -ld web

This confirms the ownership and permissions of the `web` directory:

    drwxr-xr-x 7 www-data -www-data 4096 Sep 16 22:28 web

This means:

- `www-data` owns the directory
- `www-data` is also the group
- The permissions are `drwxr-xr-x` (read/write/execute for owner, read/execute for group and others.)

To verify ownership of individual files, run:

    ls -l path/to/your/file (e.g., 'ls -l web/index.php')

This shows the file's owner and group.

{{< note >}}

If ownership is incorrect, rerun:

    sudo chown -R www-data:www-data web

And replace `www-data` with your actual web server user if different.

{{< /note >}}

#### Drupal Project Presence

> To confirm the project was scaffolded correctly, verify the presence of key files and directories:

Reviewer: ```suggestion
To confirm the project was initialized correctly, verify the presence of key files and directories:
```

    ls composer.json
    ls web/index.php
    ls -d vendor/

{{< note >}}
If any of these files are missing, Drupal has not been initialized. See the Initialize the Drupal Application Environment section to scaffold the project using Composer.
{{< /note >}}

Contributor-Safe Tips

- Be sure to consider mentioning file permissions if they are relevant to your environment (e.g., `chmod`, `chown`).

> - If `vendor/` is missing but `composer.json` is present, run `composer install`.

Reviewer: ```suggestion
- If `vendor/` is missing but `composer.json` is present, run:
 `composer install`
 - This installs required dependencies and generates the `vendor/` directory.
 - If file permissions are relevant to your environment (e.g., restrictive file creation settings or shared hosting), verify ownership and access using `chmod` and `chown`.
```

keywords: ["Drupal", "Ubuntu 22.04", "installation", "contributor-safe", "Composer"]
license: "[CC BY-ND 4.0](https://creativecommons.org/licenses/by-nd/4.0/)"
---

Drupal is a flexible content management system (CMS) designed for structured content and scalable site architecture. While it runs well in many hosting environments, Akamai’s compute instance provides a resilient foundation for high-traffic, content-rich deployments. This guide walks through installing and configuring Drupal 11 on Ubuntu 22.04 with systems-aware practices and contributor-safe steps that hold up in real-world scenarios.

## Before You Begin

This guide uses Drupal 11.1.8 to avoid known packaging issues in newer releases. Specifically a syntax error in `drupal/core-recipe-unpack` introduced in 11.2.x (see [Drupal issue #3536487](https://www.drupal.org/project/drupal/issues/3536487)). You'll install the stable version later using Composer after verifying system prerequisites.

## System Prerequisites

This guide is based on a tested configuration that ensures Drupal 11 installs cleanly and performs reliably within an Akamai-optimized infrastructure. The following prerequisites were used for validation.

> - ✅ OS: Ubuntu 22.04 LTS

Reviewer: ```suggestion
- OS: Ubuntu 22.04 LTS
```

### Create the Drupal Project Structure

> - Run the install command to generate the Drupal 11.1.8 structure. Replace `my_drupal_site` with your desired folder name. For directory layout details, see [Drupal.org's Directory Structure guide](https://www.drupal.org/docs/getting-started/understanding-drupal/directory-structure).

Author: ```suggestion
- Run the install command to generate the Drupal 11.1.8 structure. Replace `my-drupal-site` with your desired folder name. For directory layout details, see [Drupal.org's Directory Structure guide](https://www.drupal.org/docs/getting-started/understanding-drupal/directory-structure).
```
